src/ray/raylet/placement_group_resource_manager.cc

```diff
@@ -44,37 +44,65 @@ NewPlacementGroupResourceManager::NewPlacementGroupResourceManager(
       update_resources_(update_resources),
       delete_resources_(delete_resources) {}
 
-bool NewPlacementGroupResourceManager::PrepareBundle(
-    const BundleSpecification &bundle_spec) {
-  auto iter = pg_bundles_.find(bundle_spec.BundleId());
-  if (iter != pg_bundles_.end()) {
-    if (iter->second->state_ == CommitState::COMMITTED) {
-      // If the bundle state is already committed, it means that prepare request is just
-      // stale.
-      RAY_LOG(DEBUG) << "Duplicate prepare bundle request, skip it directly. This should "
-                        "only happen when GCS restarts.";
-      return true;
+bool NewPlacementGroupResourceManager::PrepareBundles(
+    const std::vector<std::shared_ptr<const BundleSpecification>> &bundle_specs) {
+  // Bundles that excluded bundles whose state is already committed.
+  std::vector<std::shared_ptr<const BundleSpecification>> preparing_bundles;
+  for (const auto &bundle_spec : bundle_specs) {
+    auto iter = pg_bundles_.find(bundle_spec->BundleId());
+    if (iter != pg_bundles_.end()) {
+      if (iter->second->state_ == CommitState::COMMITTED) {
+        // If the bundle state is already committed, it means that prepare request is just
+        // stale.
+        RAY_LOG(DEBUG)
+            << "Duplicate prepare bundle request, skip it directly. This should "
+               "only happen when GCS restarts.";
+      } else {
+        // If there was a bundle in prepare state, it already locked resources, we will
+        // return bundle resources so that we can start from the prepare phase again.
+        ReturnBundle(*bundle_spec);
+        preparing_bundles.emplace_back(bundle_spec);
+      }
     } else {
-      // If there was a bundle in prepare state, it already locked resources, we will
-      // return bundle resources so that we can start from the prepare phase again.
-      ReturnBundle(bundle_spec);
+      preparing_bundles.emplace_back(bundle_spec);
     }
   }
+  if (preparing_bundles.size() == 0) {
+    RAY_LOG(DEBUG)
+        << "All bundles have been prepared successfully before, will return directly";
+    return true;
+  }
 
-  auto resource_instances = std::make_shared<TaskResourceInstances>();
-  bool allocated = cluster_resource_scheduler_->AllocateLocalTaskResources(
-      bundle_spec.GetRequiredResources().GetResourceMap(), resource_instances);
+  std::vector<std::shared_ptr<const BundleSpecification>> prepared_bundles;
+  for (const auto &bundle_spec : preparing_bundles) {
+    auto resource_instances = std::make_shared<TaskResourceInstances>();
+    bool allocated = cluster_resource_scheduler_->AllocateLocalTaskResources(
+        bundle_spec->GetRequiredResources().GetResourceMap(), resource_instances);
+    if (!allocated) {
+      // Terminated the preparation phase and return bundles that have been successfully
+      // prepared before directly once we found one request resource failed bundle.
```

Comment:
I don't think I understand this comment. Does this mean

```
Terminate the preparation phase if any of bundle cannot be prepared
```
?

Reply:
Yes!